### PR TITLE
CHE-8510: Recognize multi-module projects

### PR DESCRIPTION
**Pull request by tolusha**
### Reference issue
https://github.com/eclipse/che/issues/8510

### CODE REVIEW

**Review by svor — APPROVED** [review]

**Review by mmorhun — APPROVED** [review]

**Review by evidolob — APPROVED** [review]

File: ide/che-core-ide-stacks/src/main/resources/stacks.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
           "machines": {
             "dev-machine": {
               "installers": [
-                "org.eclipse.che.exec", "org.eclipse.che.terminal", "org.eclipse.che.ws-agent"
+                "org.eclipse.che.exec", "org.eclipse.che.terminal", "org.eclipse.che.ws-agent", "org.eclipse.che.ls.java"
```
Comment (tsmaeder) [inline ide/che-core-ide-stacks/src/main/resources/stacks.json:31]: This turns on the java ls per default in those workpaces, correct?

File: plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/settings/compiler/JavaCompilerPreferencePresenter.java [hunk 2]
```diff
@@ -182,16 +182,6 @@ public void apply(Map<String, String> properties) throws OperationException {
                   createAndAddWidget(option);
                 }
               }
-            })
```
Comment (tsmaeder) [inline plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/settings/compiler/JavaCompilerPreferencePresenter.java:185]: Why don't we need this anymore?

File: plugins/plugin-maven/che-plugin-maven-server/src/main/java/org/eclipse/che/plugin/maven/server/core/MavenWorkspace.java [hunk 3]
```diff
@@ -177,7 +177,7 @@ private void removeProjects(List<MavenProject> removed) {
 
   public void update(List<IProject> projects) {
     manager.update(projects, true);
-    runResolve();
+    //    runResolve();
   }
 
   private void runResolve() {
```
Comment (tsmaeder) [inline plugins/plugin-maven/che-plugin-maven-server/src/main/java/org/eclipse/che/plugin/maven/server/core/MavenWorkspace.java:183]: unused?

File: wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/service/LanguageServiceUtils.java [hunk 4]
```diff
@@ -31,6 +33,14 @@ public static String removePrefixUri(String uri) {
     return uri.startsWith(FILE_PROJECTS) ? uri.substring(FILE_PROJECTS.length()) : uri;
   }
 
+  public static List<String> removePrefixUri(List<String> uris) {
+    for (int i = 0; i < uris.size(); i++) {
```
Comment (tsmaeder) [inline wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/service/LanguageServiceUtils.java]: Better to return a new copy of the list. It's not obvious for the caller that this method modifies the list that is passed in. Allocating a new List is a cheap operation. "Premature optimization is the root of all evil".

**Review by tsmaeder — CHANGES_REQUESTED** [review]

File: ide/che-core-ide-stacks/src/main/resources/stacks.json (hunk 1 above)
Comment (tolusha) [reply ide/che-core-ide-stacks/src/main/resources/stacks.json:31]: Right.

File: plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/settings/compiler/JavaCompilerPreferencePresenter.java (hunk 2 above)
Comment (tolusha) [reply plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/settings/compiler/JavaCompilerPreferencePresenter.java:185]: It annoys when jdt.ls is not ready yet. 
It is always shown at the workspace start.

File: plugins/plugin-maven/che-plugin-maven-server/src/main/java/org/eclipse/che/plugin/maven/server/core/MavenWorkspace.java (hunk 3 above)
Comment (tolusha) [reply plugins/plugin-maven/che-plugin-maven-server/src/main/java/org/eclipse/che/plugin/maven/server/core/MavenWorkspace.java:183]: Disabled not to interfere the introduced changes but can't be removed due to compilation issues.

File: wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/service/LanguageServiceUtils.java (hunk 4 above)
Comment (tolusha) [reply wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/service/LanguageServiceUtils.java]: agree

**Comment by tolusha** [comment]
@tsmaeder  It is ready to review again

**Review by tsmaeder — APPROVED** [review]

**Pull request merged**